src/llm/llmServices/utils/generationsLogger/loggerRecord.ts

```diff
@@ -0,0 +1,465 @@
+     *
+     * The reason is that, unfortunately, the current serialization-deserialization
+     * cycle neglects milliseconds.
+     */
+    readonly timestampMillis: number;
+
+    protected static floorMillisToSeconds(millis: number): number {
+        return millis - (millis % 1000);
+    }
+
+    constructor(
+        timestampMillis: number,
+        readonly modelId: string,
+        readonly responseStatus: ResponseStatus,
+        readonly choices: number,
+        readonly estimatedTokens: EstimatedTokens | undefined = undefined,
+        readonly error: LoggedError | undefined = undefined
+    ) {
+        this.timestampMillis =
+            LoggerRecord.floorMillisToSeconds(timestampMillis);
+    }
+
+    protected static readonly introLinePattern =
+        /^\[(.*)\] `(.*)` model: (.*)$/;
+
+    protected static readonly loggedErrorHeader = "! error occurred:";
+    protected static readonly loggedErrorPattern =
+        /^! error occurred: \[(.*)\] "(.*)"$/;
+
+    protected static readonly choicesPattern = /^- requested choices: (.*)$/;
+
+    protected static readonly requestTokensHeader = "- request's tokens:";
+    protected static readonly requestTokensPattern =
+        /^- request's tokens: ([0-9]+) = ([0-9]+) \(chat messages\) \+ ([0-9]+) \(max to generate\)$/;
+
+    serializeToString(): string {
+        const introInfo = this.buildStatusLine();
+        const errorInfo = this.buildErrorInfo();
+        const requestInfo = this.buildRequestInfo();
+        return `${introInfo}${errorInfo}${requestInfo}`;
+    }
+
+    toString(): string {
+        return this.serializeToString();
+    }
+
+    protected buildStatusLine(): string {
+        const timestamp = new Date(this.timestampMillis).toLocaleString();
+        return `[${timestamp}] \`${this.modelId}\` model: ${this.responseStatus}\n`;
+    }
+
+    protected buildErrorInfo(): string {
+        if (this.error === undefined) {
+            return "";
+        }
+        return `${LoggerRecord.loggedErrorHeader} [${this.error.typeName}] "${LoggerRecord.escapeNewlines(this.error.message)}"\n`;
+    }
+
+    protected buildRequestInfo(): string {
+        const choicesRequested = `- requested choices: ${this.choices}\n`;
+        const requestTokens =
+            this.estimatedTokens !== undefined
+                ? `${LoggerRecord.requestTokensHeader} ${this.estimatedTokensToString()}\n`
+                : "";
+        return `${choicesRequested}${requestTokens}`;
+    }
+
+    private estimatedTokensToString(): string {
+        return `${this.estimatedTokens!.maxTokensInTotal} = ${this.estimatedTokens!.messagesTokens} (chat messages) + ${this.estimatedTokens!.maxTokensToGenerate} (max to generate)`;
+    }
+
+    protected static escapeNewlines(text: string): string {
+        return text.replace("\n", "\\n");
+    }
+
+    static deserealizeFromString(rawRecord: string): [LoggerRecord, string] {
```

Comment:
How often is this being called? Doesn't it bring to much overhead in computational costs?

Reply:
It is a good question to be answered indeed, including the docs. I left a note at `GenerationsLogger`'s docs about the possible performance overhead and how to resolve the issue if it arises in practice one day.

No worries for now though, neither tests nor I didn't experience any problems, on the opposite, only +vibes from working with human-readable logs during debugging 🕊